flips/sync-protocol.md

```diff
@@ -0,0 +1,109 @@
+
+At any time, there is a single range of heights that the Sync Engine actively requests, which is tracked by the Sync Core. We call this the Active Range. The Active Range is parameterized by two variables `RangeStart` and `RangeEnd`, which effectively replace the `Heights` map from the existing implementation, but it can be broken up and requested by the Sync Engine in multiple segments. `RangeStart` should be greater than the local finalized block height, and `RangeEnd` should be less than or equal to the target finalized block height (more details below). The logic for updating the Active Range can be abstracted with an interface:
+
+```golang
+type ActiveRange interface {
+    // Update processes a range of blocks received from a Range Request 
+    // response and updates the requestable height range.
+    Update(headers []flow.Header, originID flow.Identifier)
+
+    // LocalFinalizedHeight is called to notify a change in the local finalized height.
+    LocalFinalizedHeight(height uint64)
+
+    // TargetFinalizedHeight is called to notify a change in the target finalized height.
+    TargetFinalizedHeight(height uint64)
+
+    // Get returns the range of requestable block heights.
+    Get() flow.Range
+}
+```
+
+There are many ways to implement this interface, but one possible approach is as follows:
+* Select values for parameters `DefaultRangeSize` and `MinResponses`
+* Let `PendingStart` be the first height greater than `LocalFinalizedHeight` that has been received less than `MinResponses` times
+* Let `RangeStart` be equal to `LocalFinalizedHeight + 1`
+* Let `RangeEnd` be the smaller of `TargetFinalizedHeight` and `PendingStart + DefaultRangeSize`
+
+The reason we keep track of `PendingStart` is to ensure that `RangeEnd` eventually increases even if the local finalized height doesn't. This is needed to address the last item in [Potential Problems](#potential-problems). 
+
+The target finalized height represents the speculated finalized block height of the overall chain, and should reflect the Sync Height Responses that have been received while accounting for the possibility that some of these responses are malicious. Therefore, the Sync Height Response processing logic should incorporate some sort of expiration / filtering mechanism. The details of this logic can be abstracted with an interface:
+
+```golang
+type TargetFinalizedHeight interface {
+    // Update processes a height received from a Sync Height Response 
+    // and updates the finalized height estimate.
+    Update(height uint64, originID flow.Identifier)
+
+    // Get returns the estimated finalized height of the overall chain.
+    Get() uint64
+}
+```
+
+One possible approach is to maintain a sliding window of the most recent Sync Height Responses, and take the median of these values. This implies that the target finalized height will always lag slightly behind the true finalized height, which may or may not be a problem depending on the block finalization rate.
```

Comment:
So far, we've mentioned about `RangeRequest`, and we'd better also talk about the `BatchRequest`. 

`BatchRequests` are independently calls from compliance engine, there is some concurrency tension between the `RangeRequest` and `BatchRequest`: when compliance engine is asking sync engine to fetch missing blocks, the sync engine might be sending range requests.

To remove the concurrency tension, we maintain a map of missing block IDs, and check the map with a time interval.

When a timer triggers, sync engine will ask sync core for RangeRequest and BatchRequest to be sent out. Now this is where we need to talk about priority. 

Should we give priority to `RangeRequest`? Yes, because if we are falling far behind, the `BatchRequest` will not help increase the local finalized height. So it's better to focus sending `RangeRequest`. For the same reason, if a missing block is far from the local finalized height, we could probably drop it, instead of keeping it in the missing block's map. That's because even if we keep them in that map, their priority is still too low to be included in a BatchRequest. We should give the `RangeRequest` priority to catch up, which is more efficient. And the `RangeRequest` would eventually reach the height of that missing block, and fetch it.

So we will store missing block IDs in a map when they are not too far from local finalized height, so that they will be included in a Batch Request.

For a RangeRequest, when there are a range of heights missing, it's easy to determine which height to fetch first: we should always start with the lowest height. But for BatchRequest, when there are a number of missing blocks with different height, and there are more blocks to be fit in a BatchRequest, how to choose them? Should we give priority to blocks with lower heights? The problem with this is that those blocks might never finalize or not exist, so we might waste effort and stuck if we keep fetching non-existing blocks from malicious nodes. Another solution is to use a _first come last serve_ rule, we will fetch those that requested recently.

Reply:
We are already giving range requests priority, so nothing changes here.

As for the prioritization of batch requests, I would consider this more of a detail, and my intention is not to prescribe specific rules about how this should be implemented. There is more than one solution here, and I don't think we have any strong reason to change from the current method, which is random.